### PR TITLE
typescript: handle constraints/bounds in generics

### PR DESCRIPTION
**Pull request by evmar**
E.g. f<T extends Foo>, we should examine Foo.  See test.

### CODE REVIEW

**Comment by evmar** [comment]
cc @ayazhafiz  (random CL)

File: kythe/typescript/testdata/generic.ts [hunk 1]
```diff
@@ -33,6 +33,14 @@ function generic<T>(x: T, y: IFace) {
   return x;
 }
 
+// Constrained generic.
+//- @IFace ref IFace
+function constrainedGeneric<T extends IFace>() {}
```
Comment (ayazhafiz) [inline kythe/typescript/testdata/generic.ts]: (nit) maybe we should define the relationship between T and IFace as well?

Comment (evmar) [reply kythe/typescript/testdata/generic.ts]: Good idea, fixed.

File: kythe/typescript/indexer.ts [hunk 2]
```diff
@@ -416,6 +416,13 @@ class Visitor {
       const kType = this.getSymbolName(sym, TSNamespace.TYPE);
       this.emitNode(kType, 'absvar');
       this.emitEdge(this.newAnchor(param.name), 'defines/binding', kType);
+      // ...<T extends A>
+      if (param.constraint) {
+        const superType = this.visitType(param.constraint);
```
Comment (jaysachs) [inline kythe/typescript/indexer.ts]: Does this handle things like "<T, K extends keyof T>" ?

(and is there only "extends" in TS, no "super" like Java has?)

And are there recursive types, e.g. <T extends Foo<T>>, and if so does this handle those?

Comment (evmar) [reply kythe/typescript/indexer.ts]: #1: yes, added a test
#2: right
#3: yes, added a test

**Comment by evmar** [comment]
PTAL, ready for submit

File: kythe/typescript/testdata/generic.ts [hunk 3]
```diff
@@ -33,6 +33,27 @@ function generic<T>(x: T, y: IFace) {
...
+//- @T defines/binding ConstrainedGenericT
+//- @IFace ref IFace
+//- ConstrainedGenericT bounded/upper IFace
+function constrainedGeneric<T extends IFace>() {}
+
+// Constrained generic with reference to type param within it.
+//- @#0T defines/binding ConstrainedGenericT2
+//- !{@#0T ref ConstrainedGenericT}
+//- @#1T ref ConstrainedGenericT2
+function constrainedGenericRef<T, K extends keyof T>() {}
```
Comment (jaysachs) [inline kythe/typescript/testdata/generic.ts]: Are there some assertions on K that could be made here?

File: kythe/typescript/testdata/generic.ts [hunk 4]
```diff
@@ -33,6 +33,27 @@ function generic<T>(x: T, y: IFace) {
...
+
+// Recursive generic type.
+//- @#0T defines/binding ConstrainedGenericT3
+//- @#1T ref ConstrainedGenericT3
+function constrainedGenericRecursive<T extends Array<T>>() {}
+
+// Default generic.
+//- @IFace ref IFace
+function defaultGeneric<T = IFace>() {}
+
```
Comment (jaysachs) [inline kythe/typescript/testdata/generic.ts]: Can you supply a default when a bound is also specified, e.g.

  function f<T extends C<T> = U>() {}

File: kythe/typescript/testdata/generic.ts [hunk 5]
```diff
@@ -33,6 +33,27 @@ function generic<T>(x: T, y: IFace) {
...
+function constrainedGenericRef<T, K extends keyof T>() {}
+
+// Recursive generic type.
+//- @#0T defines/binding ConstrainedGenericT3
+//- @#1T ref ConstrainedGenericT3
+function constrainedGenericRecursive<T extends Array<T>>() {}
+
+// Default generic.
+//- @IFace ref IFace
+function defaultGeneric<T = IFace>() {}
```
Comment (ayazhafiz) [inline kythe/typescript/testdata/generic.ts]: What is the relationship between `T` and `IFace`? Can you add a verifier goal for the binding of `T`?

File: kythe/typescript/testdata/generic.ts (hunk 3 above)
Comment (evmar) [reply kythe/typescript/testdata/generic.ts]: Added some.

File: kythe/typescript/testdata/generic.ts (hunk 4 above)
Comment (evmar) [reply kythe/typescript/testdata/generic.ts]: Yes, and I added an example, but I'm not what else to say.

File: kythe/typescript/testdata/generic.ts (hunk 5 above)
Comment (evmar) [reply kythe/typescript/testdata/generic.ts]: IFace is the default value (er, type) of T.  I added a bit but I'm really sure what to assert.

**Comment by evmar** [comment]
@kythe plz merge

**Pull request merged**